generals/gui/event_handler.py

```diff
@@ -1,77 +1,149 @@
-            # Toggle play/pause
-            case pygame.K_SPACE:
-                self.properties.paused = not self.properties.paused
-            case pygame.K_r:
-                control_events["restart"] = True
-            # Control replay frames
-            case pygame.K_h:
-                control_events["time_change"] = -1
-                self.properties.paused = True
-            case pygame.K_l:
-                control_events["time_change"] = 1
-                self.properties.paused = True
-
-
-    def __handle_mouse_click(self):
+        return (
+            x >= self.properties.display_grid_width
+            and (i + 1) * c.GUI_ROW_HEIGHT <= y < (i + 2) * c.GUI_ROW_HEIGHT
+        )
+
+    @abstractmethod
+    def handle_key_event(self, event: Event) -> Command:
+        raise NotImplementedError
+
+    @abstractmethod
+    def handle_mouse_event(self):
+        raise NotImplementedError
+
+
+class ReplayEventHandler(EventHandler):
+    def __init__(self, properties: Properties):
+        super().__init__(properties)
+        self.command = ReplayCommand()
+
+    def handle_key_event(self, event: Event) -> ReplayCommand:
+        if event.key == Q:
+            self.command.quit = True
+        elif event.key == RIGHT:
+            self.command.speed_change = 2.0
+        elif event.key == LEFT:
+            self.command.speed_change = 0.5
+        elif event.key == SPACE:
+            self.command.pause_toggle = True
+        elif event.key == R:
+            self.command.restart = True
+        elif event.key == H:
+            self.command.frame_change = -1
+        elif event.key == L:
+            self.command.frame_change = 1
+        return self.command
+
+    def handle_mouse_event(self) -> None:
         """
-        Handle mouse click event.
+        Handle mouse clicks in replay mode.
         """
         agents = self.properties.game.agents
         agent_fov = self.properties.agent_fov
 
         x, y = pygame.mouse.get_pos()
         for i, agent in enumerate(agents):
-            if self.properties.is_click_on_agents_row(x, y, i):
+            if self.is_click_on_agents_row(x, y, i):
                 agent_fov[agent] = not agent_fov[agent]
                 break
+
+
+class GameEventHandler(EventHandler):
+    def __init__(self, properties: Properties):
+        super().__init__(properties)
+        self.command = GameCommand()
+
+    def handle_key_event(self, event: Event) -> GameCommand:
+        raise NotImplementedError
+
+    def handle_mouse_event(self) -> None:
+        raise NotImplementedError
+
+
+class TrainEventHandler(EventHandler):
+    def __init__(self, properties: Properties):
+        super().__init__(properties)
+        self.command = TrainCommand()
+
+    def handle_key_event(self, event: Event) -> TrainCommand:
+        if event.key == Q:
+            self.command.quit = True
+        return self.command
+
+    def handle_mouse_event(self) -> None:
+        raise NotImplementedError
```

Comment:
Correct me if I'm wrong but isn't per Agent view also wanted in train mode?

Reply:
Yes, I missed it. Will fix tmrw